### ISSUE TITLE
[Debian] test failure on AArch64, System z and POWER8

### ISSUE THREAD

**Issue by pgrt** [persalys/persalys#748]
Hi,

I have just packaged Persalys into Debian; I am getting a failure from test cppcheck_ProbabilisticModelWindow_std on arm64 (AArch64), s390x (System z) and ppc64el (POWER8). Relevant part is

QStandardPaths: error creating runtime directory '/run/user/2952' (No such file or directory)
********* Start testing of PERSALYS::TestProbabilisticModelWindow *********
Config: Using QtTest library 5.15.4, Qt 5.15.4 (arm64-little_endian-lp64 shared (dynamic) release build; by GCC 12.1.0), debian unknown
[34m[1mWRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.[0m
PASS   : PERSALYS::TestProbabilisticModelWindow::initTestCase()
[34m[1mWRN - No main window found ! no graph settings will be displayed[0m
[34m[1mWRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.[0m
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel()
[34m[1mWRN - No main window found ! no graph settings will be displayed[0m
[34m[1mWRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.[0m
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() This plugin does not support propagateSizeHints()
FAIL!  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() 'model->getInputs()[3].getDistribution() == OT::Gamma(100, 50)' returned FALSE. (wrong input distribution)
   Loc: [./lib/test/t_ProbabilisticModelWindow_std.cxx(124)]
[34m[1mWRN - (previous message repeated 12 times)[0m
[34m[1mWRN - No main window found ! no graph settings will be displayed[0m
[34m[1mWRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.[0m
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel()
PASS   : PERSALYS::TestProbabilisticModelWindow::cleanupTestCase()
Totals: 4 passed, 1 failed, 0 skipped, 0 blacklisted, 1235ms
********* Finished testing of PERSALYS::TestProbabilisticModelWindow *********

and full build log for arm64 can be found at
  https://buildd.debian.org/status/fetch.php?pkg=persalys&arch=arm64&ver=12.0.1%2Bds-2&stamp=1660689426&raw=0

I guess the interesting part list for spotting the error is around line 124 in lib/test/t_ProbabilisticModelWindow_std.cxx

Would you have some hints or ideas for a fix? This would allow us to widen a bit the set of architectures on which Persalys can be built, currently I have it only on amd64 and mips64el.

Thanks in any case,

Best regards,

Pierre

**Comment by jschueller** [comment]
to run the gui tests you need an X server, try running ctest through  xvfb-run:
```xvfb-run -s "-screen 0 1024x768x24" ctest --output-on-failure --timeout 30 -j8```

**Comment by jschueller** [comment]
else try to insert a debug line before the QVERIFY to see the actual result:
`std::cerr <<"DEBUG:"<< model->getInputs()[3].getDistribution() << std::endl;`

**Comment by pgrt** [comment]
Hi,

Thanks for looking at my issue, sorry for having waited before testing what you suggest... Here I am.

Unfortunately, running the tests with xvfb-run did not solve the issue. I inserted the debug line as you proposed, here is the output:

********* Start testing of PERSALYS::TestProbabilisticModelWindow *********
Config: Using QtTest library 5.15.6, Qt 5.15.6 (arm64-little_endian-lp64 shared (dynamic) release build; by GCC 12.2.0), debian unknown
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
PASS   : PERSALYS::TestProbabilisticModelWindow::initTestCase()
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel()
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() This plugin does not support propagateSizeHints()
DEBUG:Gamma(k = 100, lambda = 50, gamma = -1.11022e-16)
FAIL!  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() 'model->getInputs()[3].getDistribution() == OT::Gamma(100, 50)' returned FALSE. (wrong input distribution)
   Loc: [./lib/test/t_ProbabilisticModelWindow_std.cxx(125)]
WRN - (previous message repeated 12 times)
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel()
PASS   : PERSALYS::TestProbabilisticModelWindow::cleanupTestCase()
Totals: 4 passed, 1 failed, 0 skipped, 0 blacklisted, 939ms
********* Finished testing of PERSALYS::TestProbabilisticModelWindow *********

It seems the gamma parameter in the Gamma distribution is causing the failure, although its value is close to epsilon.

Does it give you further clues?

Thanks a lot for your valuable time anyway,

Pierre

**Comment by pgrt** [comment]
Hello,

I found the above mentioned issue was linked to the companion issue
    https://github.com/openturns/openturns/issues/2179
in OpenTURNS. When applying the patch given in the above mentioned issue, the test is still failing but the output is

18/21 Test #18: cppcheck_ProbabilisticModelWindow_std .............***Failed    1.57 sec
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-pgt'
********* Start testing of PERSALYS::TestProbabilisticModelWindow *********
Config: Using QtTest library 5.15.6, Qt 5.15.6 (arm64-little_endian-lp64 shared (dynamic) release build; by GCC 12.2.0), debian unknown
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
PASS   : PERSALYS::TestProbabilisticModelWindow::initTestCase()
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestCreationFromFullModel()
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() This plugin does not support propagateSizeHints()
DEBUG: paramValueEdit->value()  is -1.11022e-16 (Debug message added by me)
FAIL!  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromWidgets() 'paramValueEdit->value() == 0' returned FALSE. (wrong gamma value)
   Loc: [./lib/test/t_ProbabilisticModelWindow_std.cxx(136)]
WRN - (previous message repeated 12 times)
WRN - No main window found ! no graph settings will be displayed
WRN - Warning! The description of the distribution ComposedCopula is [,] and cannot identify uniquely the marginal distribution. Use default description instead.
QWARN  : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel() This plugin does not support propagateSizeHints()
PASS   : PERSALYS::TestProbabilisticModelWindow::TestModificationFromModel()
PASS   : PERSALYS::TestProbabilisticModelWindow::cleanupTestCase()

(see the DEBUG line I added in this log)

I think the problem is the same as in the OpenTURNS issue, comparing double variables with == is subject to numerical accuracy concerns.
Changing line 135 in the test to
    QVERIFY2(fabs(paramValueEdit->value()) < 1.E-15, "wrong gamma value");
allows the test to pass.

Best,
Pierre
